src/executor.rs

```diff
@@ -0,0 +1,109 @@
+//! System scheduler.
+
+use heapless::binary_heap::{BinaryHeap, Max};
+
+use crate::aerugo::{error::RuntimeError, Aerugo};
+use crate::api::RuntimeApi;
+use crate::arch::Mutex;
+use crate::task::TaskStatus;
+use crate::tasklet::TaskletPtr;
+
+/// Type for the tasklet execution queue
+type TaskletQueue<const N: usize> = BinaryHeap<TaskletPtr, Max, N>;
+
+/// System scheduler.
+pub(crate) struct Executor {
+    /// Tasklet queue.
+    tasklet_queue: Mutex<TaskletQueue<8>>,
+    /// System API.
+    system_api: &'static Aerugo,
+}
+
+impl Executor {
+    /// Creates new executor instance.
+    ///
+    /// * `system_api` - System API.
+    pub(crate) const fn new(system_api: &'static Aerugo) -> Self {
+        Executor {
+            tasklet_queue: Mutex::new(BinaryHeap::new()),
+            system_api,
+        }
+    }
+
+    /// Starts tasklet scheduling.
+    pub(crate) fn run_scheduler(&self) -> ! {
+        loop {
+            self.execute_next_tasklet()
+                .expect("Failure in tasklet execution");
+        }
+    }
+
+    /// Schedules given task for execution.
+    ///
+    /// If given task is not already waiting for execution it is put to the execution queue.
+    ///
+    /// * `tasklet` - Tasklet to schedule.
+    ///
+    /// Returns `RuntimeError` in case of an error, `Ok(())` otherwise.
+    pub(crate) fn schedule_tasklet(&self, tasklet: TaskletPtr) -> Result<(), RuntimeError> {
+        let tasklet_status = tasklet.get_status();
+
+        if tasklet_status == TaskStatus::Sleeping {
+            self.add_tasklet_to_queue(tasklet)?;
+        }
+
+        Ok(())
+    }
+
+    /// Executes the next tasklet from the queue.
+    ///
+    /// Returns `RuntimeError` in case of an error, `Ok(bool)` otherwise indicating if tasklet
+    /// executed.
+    fn execute_next_tasklet(&self) -> Result<bool, RuntimeError> {
+        if let Some(tasklet) = self.get_tasklet_for_execution() {
+            tasklet.execute();
+            tasklet.set_last_execution_time(self.system_api.get_system_time());
```

Comment:
For now we only need the time when the execution ended for the scheduling purposes. In future we will implement execution monitoring and then this will be done.